cpp/src/arrow/flight/sql/odbc/odbc_api.cc

```diff
@@ -121,4 +131,121 @@ namespace arrow
     return SQL_ERROR;
   }
 
-  }  // namespace arrow
+  SQLRETURN SQLGetEnvAttr(SQLHENV env, SQLINTEGER attr, SQLPOINTER valuePtr,
+                          SQLINTEGER bufferLen, SQLINTEGER* strLenPtr) {
+    using ODBC::ODBCEnvironment;
+
+    ODBCEnvironment* environment = reinterpret_cast<ODBCEnvironment*>(env);
+
+    if (!environment) {
+      return SQL_INVALID_HANDLE;
+    }
+
+    switch (attr) {
+      case SQL_ATTR_ODBC_VERSION: {
+        if (valuePtr) {
+          SQLINTEGER* value = reinterpret_cast<SQLINTEGER*>(valuePtr);
+          *value = static_cast<SQLSMALLINT>(environment->getODBCVersion());
+          return SQL_SUCCESS;
+        } else if (strLenPtr) {
+          *strLenPtr = sizeof(SQLINTEGER);
+          return SQL_SUCCESS;
+        } else {
+          environment->GetDiagnostics().AddError(driver::odbcabstraction::DriverException(
+              "Invalid null pointer for attribute.", "HY000"));
+          return SQL_ERROR;
+        }
+      }
+
+      case SQL_ATTR_OUTPUT_NTS: {
+        if (valuePtr) {
+          // output nts always returns SQL_TRUE
+          SQLINTEGER* value = reinterpret_cast<SQLINTEGER*>(valuePtr);
+          *value = SQL_TRUE;
+          return SQL_SUCCESS;
+        } else if (strLenPtr) {
+          *strLenPtr = sizeof(SQLINTEGER);
+          return SQL_SUCCESS;
+        } else {
+          environment->GetDiagnostics().AddError(driver::odbcabstraction::DriverException(
+              "Invalid null pointer for attribute.", "HY000"));
+          return SQL_ERROR;
+        }
+      }
+
+      case SQL_ATTR_CONNECTION_POOLING:
+      case SQL_ATTR_APP_ROW_DESC: {
+        environment->GetDiagnostics().AddError(driver::odbcabstraction::DriverException(
+            "Optional feature not supported.", "HYC00"));
+        return SQL_ERROR;
+      }
+
+      default: {
+        environment->GetDiagnostics().AddError(driver::odbcabstraction::DriverException(
+            "Invalid attribute", "HYC00"));
+        return SQL_ERROR;
+      }
+    }
+
+    return SQL_ERROR;
+  }
+
+  SQLRETURN SQLSetEnvAttr(SQLHENV env, SQLINTEGER attr, SQLPOINTER valuePtr,
+                          SQLINTEGER strLen) {
+    using ODBC::ODBCEnvironment;
+
+    ODBCEnvironment* environment = reinterpret_cast<ODBCEnvironment*>(env);
+
+    if (!environment) {
+      return SQL_INVALID_HANDLE;
+    }
+
+    if (!valuePtr) {
+      environment->GetDiagnostics().AddError(
+          driver::odbcabstraction::DriverException("Invalid null pointer for attribute.", "HY024"));
+
+      return SQL_ERROR;
+    }
+
+    switch (attr) {
+      case SQL_ATTR_ODBC_VERSION: {
+        SQLINTEGER* version = reinterpret_cast<SQLINTEGER*>(valuePtr);
+        if (*version == SQL_OV_ODBC2 || *version == SQL_OV_ODBC3) {
+          environment->setODBCVersion(*version);
```

Comment:
```suggestion
        SQLINTEGER version = static_cast< SQLINTEGER >(reinterpret_cast< intptr_t >(valuePtr));
        if (version == SQL_OV_ODBC2 || version == SQL_OV_ODBC3) {
          environment->setODBCVersion(version);
```

we should use `reinterpret_cast< intptr_t >(valuePtr)` for this. 
`reinterpret_cast< intptr_t >` converts `void*` valuePtr to an integer value